File: ffb-rust/crates/ffb-mechanics/src/mixed/stats_mechanic.rs
use ffb_model::enums::PlayerStatKey;
use ffb_model::model::Game;
use crate::mechanic::{Mechanic, MechanicType};
use crate::modifiers::{InjuryContext, PlayerStatLimit, StatBasedRollModifier};
use crate::stats_drawing_modifier::StatsDrawingModifier;
use crate::stats_mechanic::StatsMechanic as StatsMechanicTrait;

/// 1:1 translation of com.fumbbl.ffb.mechanics.mixed.StatsMechanic (BB2020/BB2025).
pub struct StatsMechanic;

impl Default for StatsMechanic {
    fn default() -> Self { StatsMechanic }
}

impl Mechanic for StatsMechanic {
    fn get_type(&self) -> MechanicType { MechanicType::STAT }
}

impl StatsMechanicTrait for StatsMechanic {
    fn draw_passing(&self) -> bool { true }

    fn stat_suffix(&self) -> String { "+".to_string() }

    fn armour_is_broken(&self, armour: i32, roll: &[i32; 2], context: &InjuryContext, _game: &Game) -> bool {
        // TODO: context.armor_modifier_total(game) not yet available — using 0
        let reduced = self.reduce_armour(context, armour, 8);
        reduced <= (roll[0] + roll[1])
    }

    fn agility_modifier(&self, modifier: i32) -> StatBasedRollModifier {
        let _ = StatsDrawingModifier::positive_impairs(modifier);
        StatBasedRollModifier::new("agility", modifier)
    }

    fn improvement_increases_value(&self) -> bool { false }

    fn apply_in_game_agility_injury(&self, agility: i32, decreases: i32) -> i32 {
        agility + decreases
    }

    fn limit(&self, key: PlayerStatKey) -> PlayerStatLimit {
        match key {
            PlayerStatKey::Ma => PlayerStatLimit::new(1, 9),
            PlayerStatKey::St => PlayerStatLimit::new(1, 8),
            PlayerStatKey::Ag => PlayerStatLimit::new(1, 6),
            PlayerStatKey::Pa => PlayerStatLimit::new(1, 6),
            PlayerStatKey::Av => PlayerStatLimit::new(3, 11),
        }
    }

    fn apply_lasting_injury(&self, starting_value: i32, key: PlayerStatKey) -> i32 {
        let limit = self.limit(key);
        match key {
            PlayerStatKey::Ag | PlayerStatKey::Pa => (starting_value + 1).min(limit.get_max()),
            _ => (starting_value - 1).max(limit.get_min()),
        }
    }

    fn stat_can_be_reduced_by_injury(&self, _original_value: i32, _current_value: i32) -> bool {
        true
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ffb_model::enums::PlayerStatKey;
    use crate::stats_mechanic::StatsMechanic as StatsTrait;

    #[test]
    fn draw_passing_is_true() {
        assert!(StatsMechanic.draw_passing());
    }

    #[test]
    fn stat_suffix_is_plus() {
        assert_eq!(StatsMechanic.stat_suffix(), "+");
    }

    #[test]
    fn improvement_increases_value_is_false() {
        assert!(!StatsMechanic.improvement_increases_value());
    }

    #[test]
    fn apply_in_game_agility_injury_increments_in_mixed() {
        // mixed uses inverse agility — injury increases the stored value
        assert_eq!(StatsMechanic.apply_in_game_agility_injury(3, 1), 4);
    }

    #[test]
    fn apply_lasting_injury_increments_ag_capped_at_6() {
        assert_eq!(StatsMechanic.apply_lasting_injury(5, PlayerStatKey::Ag), 6);
        assert_eq!(StatsMechanic.apply_lasting_injury(6, PlayerStatKey::Ag), 6);
    }

    #[test]
    fn apply_lasting_injury_decrements_ma() {
        assert_eq!(StatsMechanic.apply_lasting_injury(5, PlayerStatKey::Ma), 4);
    }

    #[test]
    fn stat_can_be_reduced_always_true() {
        assert!(StatsMechanic.stat_can_be_reduced_by_injury(5, 5));
    }
}

File: ffb-java/ffb/ffb-common/src/main/java/com/fumbbl/ffb/mechanics/mixed/StatsMechanic.java
package com.fumbbl.ffb.mechanics.mixed;

import com.fumbbl.ffb.RulesCollection;
import com.fumbbl.ffb.injury.context.InjuryContext;
import com.fumbbl.ffb.mechanics.StatsDrawingModifier;
import com.fumbbl.ffb.model.Game;
import com.fumbbl.ffb.modifiers.PlayerStatKey;
import com.fumbbl.ffb.modifiers.PlayerStatLimit;

@RulesCollection(RulesCollection.Rules.BB2020)
@RulesCollection(RulesCollection.Rules.BB2025)
public class StatsMechanic extends com.fumbbl.ffb.mechanics.StatsMechanic {
	@Override
	public boolean drawPassing() {
		return true;
	}

	@Override
	public String statSuffix() {
		return "+";
	}

	@Override
	public boolean armourIsBroken(int armour, int[] roll, InjuryContext context, Game game) {
		boolean broken = (reduceArmour(context, armour, 8) <= (roll[0] + roll[1] + context.getArmorModifierTotal(game)));
		if (System.getProperty("ffb.parityDebug") != null) {
			StringBuilder mods = new StringBuilder();
			for (com.fumbbl.ffb.modifiers.ArmorModifier m : context.getArmorModifiers()) {
				mods.append(m.getName()).append(",");
			}
			System.err.println("JAVA_AVBROKE def=" + context.getDefenderId() + " armour=" + armour
				+ " reduced=" + reduceArmour(context, armour, 8) + " roll=[" + roll[0] + "," + roll[1] + "]"
				+ " modTotal=" + context.getArmorModifierTotal(game) + " mods=" + mods + " broken=" + broken);
		}
		return broken;
	}

	@Override
	public StatsDrawingModifier agilityModifier(int modifier) {
		return StatsDrawingModifier.positiveImpairs(modifier);
	}

	@Override
	public boolean improvementIncreasesValue() {
		return false;
	}

	@Override
	public int applyInGameAgilityInjury(int agility, int decreases) {
		return agility + decreases;
	}

	@Override
	public PlayerStatLimit limit(PlayerStatKey key) {
		switch (key) {
			case MA:
				return new PlayerStatLimit(1, 9);
			case ST:
				return new PlayerStatLimit(1, 8);
			case AG:
				return new PlayerStatLimit(1, 6);
			case PA:
				return new PlayerStatLimit(1, 6);
			case AV:
				return new PlayerStatLimit(3, 11);
			default:
				return new PlayerStatLimit(0, 0);
		}
	}

	@Override
	public int applyLastingInjury(int startingValue, PlayerStatKey key) {
		PlayerStatLimit limit = limit(key);

		switch (key) {
			case AG:
			case PA:
				return Math.min(startingValue + 1, limit.getMax());
			default:
				return Math.max(startingValue - 1, limit.getMin());
		}
	}

	@Override
	public boolean statCanBeReducedByInjury(int originalValue, int currentValue) {
		return true;
	}
}
